src/mds/CInode.cc

```diff
@@ -5492,6 +5512,59 @@ void CInode::setxattr_ephemeral_dist(bool val)
   _get_projected_inode()->set_ephemeral_distributed_pin(val);
 }
 
+std::string CInode::get_bal_rank_mask_from_xattrs(bool projected_node)
+{
+  const auto& pxattrs = (projected_node && !projected_nodes.empty())
+                        ? projected_nodes.front().xattrs : get_xattrs();
+  if (pxattrs) {
+    auto it = pxattrs->find("ceph.dir.bal.mask");
+    if (it != pxattrs->end()) {
+        std::string val(it->second.c_str(), it->second.length());
+        return val;
+    }
+  }
+  return "";
+}
+
+CInode *CInode::get_rank_mask_inode(bool inherit)
+{
+  if (!g_conf().get_val<bool>("mds_bal_export_pin"))
```

Comment:
This will be too expensive to run so frequently. Please cache the config variable in MDCache (as we do elsewhere).

Reply:
@batrick I fixed it.